src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+
+static ProfilerRegion pr_operation("Replay Operation");
+static ProfilerRegion pr_kernel_launch("KernelLaunch");
+static ProfilerRegion pr_barrier("Barrier");
+static ProfilerRegion pr_memset_async("MemsetAsync");
+static ProfilerRegion pr_reduce_expanded("ReduceExpanded");
+static ProfilerRegion pr_expand("Expand");
+static ProfilerRegion pr_compress("Compress");
+static ProfilerRegion pr_memcpy_async("MemcpyAsync");
+static ProfilerRegion pr_mkperm("Mkperm");
+static ProfilerRegion pr_block_reduce("BlockReduce");
+static ProfilerRegion pr_block_prefix_reduce("BlockPrefixReduce");
+static ProfilerRegion pr_aggregate("Aggregate");
+static ProfilerRegion pr_free("Free");
+static ProfilerRegion pr_output("Output");
+
+int Recording::replay(const uint32_t *replay_inputs, uint32_t *outputs) {
+
+    uint32_t n_kernels = 0;
+
+    ThreadState *ts = thread_state(backend);
+    if (dynamic_cast<RecordThreadState *>(ts) != nullptr)
+        jitc_raise("replay(): Tried to replay while recording!");
+
+    OptixShaderBindingTable *tmp_sbt = ts->optix_sbt;
+    scoped_set_context_maybe guard2(ts->context);
+
+    replay_variables.clear();
+
+    replay_variables.reserve(this->record_variables.size());
+    for (RecordVariable &rv : this->record_variables) {
+        replay_variables.push_back(ReplayVariable(rv));
+    }
+
+    jitc_log(LogLevel::Info, "replay(): inputs");
+    // Populate with input variables
+    for (uint32_t i = 0; i < this->inputs.size(); ++i) {
+        Variable *input_variable = jitc_var(replay_inputs[i]);
+        replay_variables[this->inputs[i]].init_from_input(input_variable);
+        jitc_log(LogLevel::Debug, "    input %u: r%u maps to slot s%u", i,
+                 replay_inputs[i], this->inputs[i]);
+    }
+
+    // The main loop that executes each operation
+    for (uint32_t i = 0; i < this->operations.size(); ++i) {
+        Operation &op = this->operations[i];
+        ProfilerPhase profiler(pr_operation);
+        if (!op.enabled)
+            continue;
+
+        switch (op.type) {
+            case OpType::KernelLaunch: {
+                ProfilerPhase profiler(pr_kernel_launch);
+
+                // Test if this kernel is still in the cache
+                auto it = state.kernel_cache.find(
+                    *op.kernel.key,
+                    KernelHash::compute_hash(op.kernel.hash.high64,
+                                             op.kernel.key->device,
+                                             op.kernel.key->flags));
+                if (it == state.kernel_cache.end())
+                    jitc_raise(
+                        "replay(): Freezing functions relies on the kernel "
+                        "cache. It looks like this has been cleared since "
+                        "recording the function.");
+
+                // Reconstruct the \ref kernel_params for this launch given the
+                // allocations when replaying.
+                kernel_params.clear();
```

Comment:
Is there any risk this interferes with an unrelated pending kernel launch?

Reply:
That's not the same `kernel_params `as in `eval.cpp`, as I didn't want to expose that in `eval.h` it should not clear that one. Though we could make it the same vector, since we cannot be in `jitc_eval` and `jitc_freeze_replay` at the same time due to the `state.lock` mutex.